### ISSUE TITLE
storage: long time spent in evalGC

### ISSUE THREAD

**Issue by tschottdorf** [cockroachdb/cockroach#16068]
On our internal test cluster `adriatic`, we see GC requests take between 7 and 17 seconds. This is under `kv` load (`--concurrency 64 --tolerate-errors --min-block-bytes=64 --max-block-bytes=128 --cycle-length=1000000`), but when the load is stopped and GC continues to happen, the latencies look about the same (even though the GC operations at that point were essentially alone).

Adding additional instrumentation revealed that `evalGC` took up the bulk of the time. The main loop, iterating over ~7k keys in one example, handled approximately 1000 keys per second, where each key (empirically) has ~2 versions (I see two calls to `iter.Next()` most of the time).

When the GC queue was preparing the GCRequest (whose evaluation takes a long time as discussed above), that is usually done in <0.5s, though it scans a lot more data.

Unfortunately, the process crashes when pprof'ed. I'll have to look into that as well.

Here's a typical trace (with write load):

```

0s
[gc,n3,s3,r175/5:/Table/51/1/54{02980…-77157…}] processing replica with score queue=true with 2.02/fuzz(1.06)=1.91=valScaleScore(13.38)*deadFrac(0.14)+intentScore(0.00) likely last GC: 1h2m3.408221298s ago, 1.7 MiB non-live, curr. age 13 GiB*s, min exp. reduction: 12 GiB*s
0s
[gc,n3,s3,r175/5:/Table/51/1/54{02980…-77157…}] iterating through range
0.35s
[gc,n3,s3,r175/5:/Table/51/1/54{02980…-77157…}] pushing up to 0 transactions (concurrency 25)
0.35s
[gc,n3,s3,r175/5:/Table/51/1/54{02980…-77157…}] resolving up to 0 intents
0.35s
[gc,n3,s3,r175/5:/Table/51/1/54{02980…-77157…}] processing abort cache
0.35s
[gc,n3,s3,r175/5:/Table/51/1/54{02980…-77157…}] MVCC stats: {ContainsEstimates:false LastUpdateNanos:1495483552570672719 IntentAge:0 GCBytesAge:14064197957 LiveBytes:10538854 LiveCount:74475 KeyBytes:2045427 KeyCount:74475 ValBytes:10245161 ValCount:89771 IntentBytes:0 IntentCount:0 SysBytes:855 SysCount:9}
0.35s
[gc,n3,s3,r175/5:/Table/51/1/54{02980…-77157…}] assembled GC keys, now proceeding to GC; stats so far {Now:1495483552.305844294,106 Policy:{TTLSeconds:600} NumKeysAffected:12565 IntentsConsidered:0 IntentTxns:0 TransactionSpanTotal:0 TransactionSpanGCAborted:0 TransactionSpanGCCommitted:0 TransactionSpanGCPending:0 TxnSpanGCThreshold:1495479952.305844294,106 AbortSpanTotal:0 AbortSpanConsidered:0 AbortSpanGCNum:0 PushTxn:0 ResolveTotal:0 ResolveSuccess:0 Threshold:1495482952.305844294,0}
0.35s
[gc,n3,s3,r175/5:/Table/51/1/54{02980…-77157…}] read-write path
0.35s
[gc,n3,s3,r175/5:/Table/51/1/54{02980…-77157…}] command queue
0.36s
[gc,n3,s3,r175/5:/Table/51/1/54{02980…-77157…}] raft
0.37s
[gc,n3,s3,r175/5:/Table/51/1/54{02980…-77157…}] starting GC evaluation for 12565 keys
23.4s
[gc,n3,s3,r175/5:/Table/51/1/54{02980…-77157…}] done with GC evaluation at 546.27 keys/sec
23.4s
[gc,n3,s3,r175/5:/Table/51/1/54{02980…-77157…}] evaluated request
23.4s
[gc,n3,s3,r175/5:/Table/51/1/54{02980…-77157…}] acquired {raft,replica}mu
23.4s
[gc,n3,s3,r175/5:/Table/51/1/54{02980…-77157…}] proposing at MaxLeaseIndex=3312801 (~0 requests in-flight)
23.4s
[gc,n3,s3,r175/5:/Table/51/1/54{02980…-77157…}] applying command
23.4s
[gc,n3,s3,r175/5:/Table/51/1/54{02980…-77157…}] done GC'ing, new score is queue=false with 0.23/fuzz(1.06)=0.22=valScaleScore(8.35)*deadFrac(0.03)+intentScore(0.00) likely last GC: 23.407729172s ago, 275 KiB non-live, curr. age 1.3 GiB*s, min exp. reduction: 1.2 GiB*s
```

I did check that `evalGC` in fact loses time fairly evenly as it goes through the keys.

**Comment by petermattis** [comment]
Wow, that's slow. I took a quick glance at `MVCCGarbageCollect` (I assume that's the slow operation here) and nothing jumped out.

PS The `raft` event is misplaced with proposer-evaluated-kv.

**Comment by tschottdorf** [comment]
Yes, MVCCGarbageCollect is the slow one, and I also thought it looked reasonable.

**Comment by tschottdorf** [comment]
I have stopped the load generators a few minutes back but boy, these results are just getting worse:

```
0.02m
[gc,n6,s6,r201/2:/Table/51/1/-{615153…-538927…}] starting GC evaluation for 17753 keys
1.02m
[gc,n6,s6,r201/2:/Table/51/1/-{615153…-538927…}] done with GC evaluation at 295.78 keys/sec
```

All the while, the machine seems pretty bored, though the GC frequency is perhaps a little high?

```
[n1] runtime stats: 256 MiB RSS, 206 goroutines, 39 MiB/46 MiB/100 MiB GO alloc/idle/total, 60 MiB/91 MiB CGO alloc/total, 957.28cgo/sec, 0.34/0.04 %(u/s)time, 0.00 %gc (11x)
```

**Comment by bdarnell** [comment]
Are the input keys to `GC` sorted? It looks like the answer is yes (or at least they're in a small number of sorted chunks), but I haven't traced all the possibilities. If the keys were unsorted (e.g. because they passed through a map somewhere along the way), that would really slow `MVCCGarbageCollect` down. 

`MVCCGarbageCollect` is using a non-prefix iterator, so it can't use bloom filters and must seek all levels of the LSM. Since the keys are likely to be fairly sparse, this may be worse than using a prefix iterator for each key.

**Comment by tschottdorf** [comment]
I also thought of (parts of) that, but the keys should be mostly sorted. I'll add some more instrumentation and try using a prefix iterator.

**Comment by tschottdorf** [comment]
```
1.05s
[gc,n2,s2,r159/2:/Table/51/1/-4{54028…-46807…}] starting GC evaluation for 16515 keys
54s
[gc,n2,s2,r159/2:/Table/51/1/-4{54028…-46807…}] done with GC evaluation at 311.77 keys/sec
```

with the following patch

```diff
diff --git a/pkg/storage/engine/mvcc.go b/pkg/storage/engine/mvcc.go
index 6fcb44db4..2db081276 100644
--- a/pkg/storage/engine/mvcc.go
+++ b/pkg/storage/engine/mvcc.go
@@ -22,6 +22,7 @@ import (
 	"fmt"
 	"math"
 	"sync"
+	"time"
 
 	"golang.org/x/net/context"
 
@@ -2111,11 +2112,32 @@ func MVCCGarbageCollect(
 	keys []roachpb.GCRequest_GCKey,
 	timestamp hlc.Timestamp,
 ) error {
-	iter := engine.NewIterator(false)
-	defer iter.Close()
+
 	// Iterate through specified GC keys.
 	meta := &enginepb.MVCCMetadata{}
+	log.Eventf(ctx, "starting GC evaluation for %d keys", len(keys))
+	begin := time.Now()
+	defer func() {
+		log.Eventf(ctx, "done with GC evaluation at %.2f keys/sec", float64(len(keys))*1E9/float64(time.Now().Sub(begin)))
+	}()
+
+	var iter Iterator
+	if len(keys) > 0 {
+		defer func() {
+			iter.Close()
+		}()
+	}
+	var lastKey roachpb.Key
 	for _, gcKey := range keys {
+		if iter != nil {
+			iter.Close()
+		}
+		if bytes.Compare(lastKey, gcKey.Key) >= 0 {
+			log.Event(ctx, "current key sorts in front of previous key")
+		}
+		lastKey = gcKey.Key
+		iter = engine.NewIterator(true)
+
 		encKey := MakeMVCCMetadataKey(gcKey.Key)
 		ok, metaKeySize, metaValSize, err := mvccGetMetadata(iter, encKey, meta)
 		if err != nil {
@@ -2166,6 +2188,7 @@ func MVCCGarbageCollect(
 			} else if !ok {
 				break
 			}
+
 			unsafeIterKey := iter.UnsafeKey()
 			if !unsafeIterKey.Key.Equal(encKey.Key) {
 				break

```

Granted this is under load, but it seems that the keys/sec is getting lower and lower. Is this another tombstone issue? Do we have a reference for tapping into the underlying RocksDB stats associated to the iterator(s)?

**Comment by petermattis** [comment]
The prefix iter can be reused between loop iterations since we're always seeking as the first step. 

> Is this another tombstone issue? Do we have a reference for tapping into the underlying RocksDB stats associated to the iterator(s)?

I don't recall a stat which indicated excessive tombstones to skip. You could try shutting down all of the nodes and forcing a full compaction. I'm pretty sure there is a `cockroach debug` command to do so. That would eliminate all of the tombstones.

**Comment by tschottdorf** [comment]
Reusing the iter didn't fix this. I'm sure I can repro this badness locally. Will prepare something.

My RocksDB question was archaeological: when we discovered the first tombstone issue, I think you made use of some RocksDB debugging facilities to track it down. But maybe I'm also just misremembering this; either way, I'm not quite there yet.

**Comment by petermattis** [comment]
For the first tombstone issue, I was able to create a test which replicated it. With the test in hand I was able to add instrumentation first to the Go code and eventually to RocksDB that indicated where the slowness was coming from. I don't recall there being a metric or anything which would have highlighted the problem, but I might be misremembering.

**Comment by tschottdorf** [comment]
```
./cockroach debug sstables /mnt/data1/
0 [   1M  1 ]: 1M
4 [   5M  3 ]: 4M 953K 666K
5 [ 232M 21 ]: 27M 25M 22M 20M 17M[2] 15M[2] 12M[2] 8M[2] 6M[3] 5M 4M[2] 3M 1M 160K
6 [   2G 46 ]: 100M[21] 99M 86M 60M 19M 12M 5M[3] 4M 3M 2M 1021K 947K 907K 454K 404K 237K 113K 110K 82K 80K 70K 53K 24K 11K
```

Looks OK to me (at least it's not so outrageous that *I* would spot it).

Compacted all stores, but...

```
0.76s
[gc,n5,s5,r139/3:/Table/51/1/87{28163…-95870…}] starting GC evaluation for 21258 keys
52.3s
[gc,n5,s5,r139/3:/Table/51/1/87{28163…-95870…}] done with GC evaluation at 412.68 keys/sec
52.3s
[gc,n5,s5,r139/3:/Table/51/1/87{28163…-95870…}] evaluated request
```

I'll make a local data set and get to benchmarking.

**Comment by petermattis** [comment]
Ok, so probably not a tombstone issue. I wonder if a benchmark for `MVCCGarbageCollect` would show the bad behavior.

**Comment by tschottdorf** [comment]
I'll post the benchmark shortly, but so far the answer to that is no:

`adriatic` is running kv with a cycle-length of 1 million. It gets less than 2000 writes/sec.

So: 1 million keys / (2000 writes/sec) = 500sec, i.e. we get one new version every 500 sec (let's make that 600sec=10 min for easier math).

GC happens approximately every hour in the traces I've seen, so you expect 6 writes to each key GC'ed, and certainly in the worst case you have to delete all of them. The traces I've seen take around half minute have on the order of 15.000 keys, and rarely process keys at a rate faster than 500 keys/sec.

So what I benchmarked is: write 15.000 keys with 6 versions each, and run a GCRequest that GCs the five non-live versions of all of these keys. The result: around 500ms, or ~30.000keys/sec.

Sure, this is in-mem and the writes happen just before the reads and in a certain order, and perhaps I messed up the benchmark, but I did check that `engine.Clear` shows up in the profile, and it does, and takes approximately 4-5x the time we spend in `iter.Next()` and `mvccGetMetadata`, respectively.

**Comment by tschottdorf** [comment]
The good news is that I now have a dataset that reproduces this badness locally:

```
13:09:32.485104	 .   337	... [gc,n1,s1,r12/1:/{Table/51-Max}] starting GC evaluation for 1001 keys
13:10:01.478282	28.993178	... [gc,n1,s1,r12/1:/{Table/51-Max}] current key sorts in front of previous key
13:10:01.478347	 .    64	... [gc,n1,s1,r12/1:/{Table/51-Max}] done with GC evaluation at 34.53 keys/sec
```

**Comment by tschottdorf** [comment]
Ok, new theory thanks to jerry-rig debugging. My math above is correct but implicitly assumes that GC *keeps up* with the writes. However, GC on adriatic *did not keep up* with the writes. Correspondingly, I have reason to assume that there weren't ~6 old versions, but rather ~1000 and growing. Benchmarking a key with ~1000 old versions gives me numbers that would end up in the tens of seconds when there are ~10k+ keys. That all seems reasonable. I'm going to have to verify that that's the case on adriatic (I came up with this now-obvious theory with my local dataset, which had a different write pattern - much more focused, so many more versions per key), but I already have the code to validate, so it shouldn't take long.

@petermattis we could just use `ClearRange` to kill all of the GC'able versions of a key wholesale (instead of iterating and individual `engine.Clear`), correct?

**Comment by petermattis** [comment]
Adding too many range tombstones can slow down reads. I don't know the precise trade-off where it is worthwhile, but using range tombstones to delete old versions sounds problematic.

Also, we'd still have to iterate over the old versions in order to update the MVCC stats correctly. I'm a bit surprised that the calls to `Clear` are so problematic.

**Comment by bdarnell** [comment]
Does this mean we need another source of backpressure/flow control, to slow down writes when the existing range contains too much garbage? But this seems tricky to do correctly since garbage stays there for a long time, in contrast to things like the raft log. Maybe we need to report on this and encourage admins of high-churn clusters to reduce their GC TTL?

Using ClearRange instead of Clear should be fine, at least from a correctness perspective. Not sure what it does for performance.

**Comment by tschottdorf** [comment]
Honestly, I think we can do so much better with the GCQueue compared to what it is now. I remember how the initial design didn't have explicit replicated GC but simply threw out values that were no longer visible - I think we abandoned that in an attempt to make replica divergence easier to check. However, I think we should sit down and reconsider whether that attempt wouldn't make more sense. We are already making sure that GC'ed data can't be read (the GC threshold), and if we exclude GCable versions from consistency checks those would continue working as expected. Bonus is we can omit this data from snapshots as well. But I'll file follow-ups for these investigations.

On adriatic, looks like something else is going on. I'm still seeing the slow GC evaluations, but the numbers aren't as clear cut. Will look more after lunch.

**Comment by tschottdorf** [comment]
Analysis of the sample trace below (which has hacky profiling added): The bulk of the ~34s of evaluation for 115171 keys is spent in these blocks outlined below. However, each block individually per iteration takes about the same amount of time: 250-600µs. I take it that those individual operation times are much slower than one should reasonably expect, after all, my smallest benchmark in #16100 took only 22.6us. There's so much going on on these machines that it's unfortunately hard to really get an idea of what's causing this slowness, though. For example, one particularly long GC run coincided with a flood of ~0.6s handleRaftReady calls. Whether it caused them or the other way around, I'm not sure.

~9s for loading metadata (time spent for 1: 580.82µs/op (8.811624136s total, 15171 count)

```
		if bytes.Compare(lastKey, gcKey.Key) >= 0 {
			log.Event(ctx, "current key sorts in front of previous key")
		}
		lastKey = gcKey.Key

		encKey := MakeMVCCMetadataKey(gcKey.Key)
		ok, metaKeySize, metaValSize, err := mvccGetMetadata(iter, encKey, meta)
		if err != nil {
			return err
		}
		if !ok {
			continue
		}
		inlinedValue := meta.IsInline()
		implicitMeta := iter.UnsafeKey().IsValue()
```

~16.5s for iterating through key versions: (pretty exactly 3 per key)

```go
		if !implicitMeta {
			// The iter is pointing at an MVCCMetadata, advance to the next entry.
			iter.Next()
		}

		// Now, iterate through all values, GC'ing ones which have expired.
		for ; ; iter.Next() {
```

~7.5 for `engine.Clear` (at most 2x#keys times)

```go
			if !gcKey.Timestamp.Less(unsafeIterKey.Timestamp) {
				if ms != nil {
					ms.Add(updateStatsOnGC(gcKey.Key, mvccVersionTimestampSize,
						int64(len(iter.UnsafeValue())), nil, unsafeIterKey.Timestamp.WallTime,
						timestamp.WallTime))
				}
				if err := engine.Clear(unsafeIterKey); err != nil {
					return err
				}
			}
```

--

```
0s
[gc,n3,s3,r102/4:/Table/51/1/-5{51905…-44774…}] processing replica with score queue=true with 2.01/fuzz(1.22)=1.65=valScaleScore(17.06)*deadFrac(0.10)+intentScore(0.00) likely last GC: 1h57m58.318952364s ago, 2.0 MiB non-live, curr. age 20 GiB*s, min exp. reduction: 19 GiB*s
0s
[gc,n3,s3,r102/4:/Table/51/1/-5{51905…-44774…}] iterating through range
0.88s
[gc,n3,s3,r102/4:/Table/51/1/-5{51905…-44774…}] pushing up to 0 transactions (concurrency 25)
0.88s
[gc,n3,s3,r102/4:/Table/51/1/-5{51905…-44774…}] resolving up to 0 intents
0.88s
[gc,n3,s3,r102/4:/Table/51/1/-5{51905…-44774…}] processing abort cache
0.88s
[gc,n3,s3,r102/4:/Table/51/1/-5{51905…-44774…}] MVCC stats: {ContainsEstimates:false LastUpdateNanos:1495667005555070462 IntentAge:0 GCBytesAge:21728775366 LiveBytes:19841890 LiveCount:147295 KeyBytes:3904675 KeyCount:147295 ValBytes:18060262 ValCount:165820 IntentBytes:0 IntentCount:0 SysBytes:647 SysCount:9}
0.88s
[gc,n3,s3,r102/4:/Table/51/1/-5{51905…-44774…}] assembled GC keys, now proceeding to GC; stats so far {Now:1495667004.994423667,0 Policy:{TTLSeconds:600} NumKeysAffected:15171 IntentsConsidered:0 IntentTxns:0 TransactionSpanTotal:0 TransactionSpanGCAborted:0 TransactionSpanGCCommitted:0 TransactionSpanGCPending:0 TxnSpanGCThreshold:1495663404.994423667,0 AbortSpanTotal:0 AbortSpanConsidered:0 AbortSpanGCNum:0 PushTxn:0 ResolveTotal:0 ResolveSuccess:0 Threshold:1495666404.994423667,0}
0.88s
[gc,n3,s3,r102/4:/Table/51/1/-5{51905…-44774…}] read-write path
0.88s
[gc,n3,s3,r102/4:/Table/51/1/-5{51905…-44774…}] command queue
0.89s
[gc,n3,s3,r102/4:/Table/51/1/-5{51905…-44774…}] raft
0.9s
[gc,n3,s3,r102/4:/Table/51/1/-5{51905…-44774…}] starting GC evaluation for 15171 keys
33.9s
[gc,n3,s3,r102/4:/Table/51/1/-5{51905…-44774…}] time spent for 0: 376ns/op (5.716568ms total, 15171 count)
33.9s
[gc,n3,s3,r102/4:/Table/51/1/-5{51905…-44774…}] time spent for 1: 580.82µs/op (8.811624136s total, 15171 count)
33.9s
[gc,n3,s3,r102/4:/Table/51/1/-5{51905…-44774…}] time spent for 2: 289ns/op (4.386647ms total, 15171 count)
33.9s
[gc,n3,s3,r102/4:/Table/51/1/-5{51905…-44774…}] time spent for 3: 363.815µs/op (16.558352101s total, 45513 count)
33.9s
[gc,n3,s3,r102/4:/Table/51/1/-5{51905…-44774…}] time spent for 4: 517ns/op (15.69258ms total, 30342 count)
33.9s
[gc,n3,s3,r102/4:/Table/51/1/-5{51905…-44774…}] time spent for 5: 362ns/op (10.998063ms total, 30342 count)
33.9s
[gc,n3,s3,r102/4:/Table/51/1/-5{51905…-44774…}] time spent for 6: 249.921µs/op (7.583115116s total, 30342 count)
33.9s
[gc,n3,s3,r102/4:/Table/51/1/-5{51905…-44774…}] time spent for 7: 602ns/op (9.140031ms total, 15171 count)
33.9s
[gc,n3,s3,r102/4:/Table/51/1/-5{51905…-44774…}] done with GC evaluation at 459.74 keys/sec
33.9s
[gc,n3,s3,r102/4:/Table/51/1/-5{51905…-44774…}] evaluated request
33.9s
[gc,n3,s3,r102/4:/Table/51/1/-5{51905…-44774…}] acquired {raft,replica}mu
33.9s
[gc,n3,s3,r102/4:/Table/51/1/-5{51905…-44774…}] proposing at MaxLeaseIndex=3708177 (~0 requests in-flight)
33.9s
[gc,n3,s3,r102/4:/Table/51/1/-5{51905…-44774…}] applying command
34.1s
[gc,n3,s3,r102/4:/Table/51/1/-5{51905…-44774…}] done GC'ing, new score is queue=false with 0.33/fuzz(1.22)=0.27=valScaleScore(14.39)*deadFrac(0.02)+intentScore(0.00) likely last GC: 34.057162741s ago, 376 KiB non-live, curr. age 3.1 GiB*s, min exp. reduction: 2.9 GiB*s
```

**Comment by tschottdorf** [comment]
I added some hacks to avoid GCRequests blocking other requests, and now (I think for the first time) I can see some evaluations that are pretty fast:

```
[gc,n5,s5,r57/3:/Table/51/1/-{819307…-751915…}] done with GC evaluation at 12929.20 keys/sec
```

The log is still full of slow readys and timed out requests. I'll reduce the load on the cluster so that it's not constantly overwhelmed. Perhaps chasing the GC queue is wasted time here on a cluster that busy.

**Comment by tschottdorf** [comment]
Ha! I think I may have accidentally discovered something interesting. The "hack" mentioned above includes this:

```
func (ss *SpanSet) checkAllowed(access SpanAccess, span roachpb.Span) error {
	return nil // HACK
```

and I think that makes all the difference, because of this code:

```
	batch := r.store.Engine().NewBatch()
	if spans != nil {
		batch = makeSpanSetBatch(batch, spans)
	}
```

Having made `checkAllowed` a noop is what makes things benchmark-fast.

![image](https://cloud.githubusercontent.com/assets/5076964/26432623/0d7c13b4-40cb-11e7-887f-33d08f3da5e0.png)

Should be straightforward to find out why `checkAllowed` is so slow.

**Comment by bdarnell** [comment]
Ah, debug assertions in production strike again! `checkAllowed` was meant to be turned off in production once propEvalKV was complete. It's quite inefficient (could be made better, but probably not worth it unless we want to keep it on in prod).

**Comment by tschottdorf** [comment]
@bdarnell I'll take a look at removing them tomorrow. In the process of poking around however, I also found some oddities with regards of our handling of the `GCThreshold` key - I was under the assumption that all reads would have that in their SpanSet, but it seems that they don't (though the key or rather its in-memory state version are checked before going in for reads). That seems like it opens up a race between GC and reads (not for those with BeginTransaction, as that overlaps GCRequest in the handling of TxnSpanThreshold). Similarly, `GCRequest` doesn't declare the abort span from which it may delete. Perhaps you remember whether there's something to dispel my concerns, but I think we're just being sloppy and getting away with it (as it's pretty impossible in practice to hit these anomalies). I think we should tackle #7880 and clean these up along the way.

**Comment by petermattis** [comment]
Nice catch! I knew that `checkAllowed` was still on, but hadn't measured a perf benefit to disabling it. I should have noticed that it is O(n) in the number of keys in the request.

**Comment by tschottdorf** [comment]
Fixed in #16131.
